[dynamo] Add ignore_fresh_unbacked_symbols for foreach ops with scalar values

mlazos opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #170289
* __->__ #170288

When capture_scalar_outputs is enabled and a tensor scalar is passed to
foreach ops like _foreach_addcmul, the .item() call creates an unbacked
symbol that is not tracked in output tensor shapes. This causes
PendingUnbackedSymbolNotFound errors.

Fix by wrapping the wrap_fx_proxy call with ignore_fresh_unbacked_symbols
for ops that consume scalar values for computation only (not for output shapes).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

mlazos committed:
[dynamo] Add ignore_fresh_unbacked_symbols for foreach ops with scalar values

When capture_scalar_outputs is enabled and a tensor scalar is passed to
foreach ops like _foreach_addcmul, the .item() call creates an unbacked
symbol that is not tracked in output tensor shapes. This causes
PendingUnbackedSymbolNotFound errors.

Fix by wrapping the wrap_fx_proxy call with ignore_fresh_unbacked_symbols
for ops that consume scalar values for computation only (not for output shapes).

[ghstack-poisoned]

mlazos committed:
Update on "[dynamo] Add ignore_fresh_unbacked_symbols for foreach ops with scalar values"


When capture_scalar_outputs is enabled and a tensor scalar is passed to
foreach ops like _foreach_addcmul, the .item() call creates an unbacked
symbol that is not tracked in output tensor shapes. This causes
PendingUnbackedSymbolNotFound errors.

Fix by wrapping the wrap_fx_proxy call with ignore_fresh_unbacked_symbols
for ops that consume scalar values for computation only (not for output shapes).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

mlazos committed:
Update on "[dynamo] Add ignore_fresh_unbacked_symbols for foreach ops with scalar values"


When capture_scalar_outputs is enabled and a tensor scalar is passed to
foreach ops like _foreach_addcmul, the .item() call creates an unbacked
symbol that is not tracked in output tensor shapes. This causes
PendingUnbackedSymbolNotFound errors.

Fix by wrapping the wrap_fx_proxy call with ignore_fresh_unbacked_symbols
for ops that consume scalar values for computation only (not for output shapes).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

aorenste commented:
Is the test correct? I was curious to see the actual symbol creation location and copied the test locally and it passed without the other changes...

mlazos committed:
Update on "[dynamo] Add ignore_fresh_unbacked_symbols for foreach ops with scalar values"


When capture_scalar_outputs is enabled and a tensor scalar is passed to
foreach ops like _foreach_addcmul, the .item() call creates an unbacked
symbol that is not tracked in output tensor shapes. This causes
PendingUnbackedSymbolNotFound errors.

Fix by wrapping the wrap_fx_proxy call with ignore_fresh_unbacked_symbols
for ops that consume scalar values for computation only (not for output shapes).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

mlazos commented:
> Is the test correct? I was curious to see the actual symbol creation location and copied the test locally and it passed without the other changes...

lol, I needed it to be 1 - val. This adds even more to the mystery. I also ran into another issue when having raw val as a non-kwarg of the foreach_pow, about a guard, I can file two issues for followup. 

@aorenste thanks for pointing this out and trying it!

williamwen42 reviewed (approved):
Approving to unblock (given CI is green) - @laithsakka @bobrenjc93 @pianpwk if there are issues, we can follow up?

pianpwk reviewed (approved):
it seems like we should to fix the C++ foreach handling to not call item/local_scalar_dense, but I couldn't figure out where that was happening..
